### PR TITLE
Support correlated scalar subqueries (v6)

### PR DESCRIPTION
**Pull request by kokosing**
Support correlated count scalar subqueries

A scalar subquery is a simple query which aggregates all the rows into a
single group (empty GROUP BY clause). For example:

SELECT min(a) FROM t WHERE b > 0
This commit add support for correlated count scalar subqueries which
have WHERE cluase of conjunction of equality comparison. For example:

SELECT (SELECT min(a) FROM t1(a,c) WHERE b = 0 AND c = d) FROM t2(b,d)
Logically such query is rewritten into left outer join where correlation
used in eqality comparisons (from WHERE cluase) are used as join
condition.

### CODE REVIEW

**Comment by kokosing** [comment]
Supersedes #6035

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java [hunk 1]
```diff
@@ -32,8 +32,8 @@
 public class AssignUniqueIdOperator
         implements Operator
 {
-    private static final long ROW_IDS_PER_REQUEST = 1 << 20;
-    private static final long MAX_ROW_ID = 1 << 40;
+    private static final long ROW_IDS_PER_REQUEST = 1L << 20L;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: This should go in the commit that introduced the new node

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Scalar aggregation is aggregation with GROUP BY 'a constant' (or empty GROUP BY).
+ * It always returns single row in Presto.
+ * <p>
+ * This optimizer can rewrite correlated scalar aggregation subquery to left outer join in a way described here:
+ * https://docs.google.com/document/d/18HN7peS2eR8lZsErqcmnoWyMEPb6p4OQeidH1JP_EkA/edit#heading=h.hb6ihepq7q74
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: This should link to the wiki document

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+ * <p>
+ * From:
+ * - Apply (with correlation list: [C])
+ *   - (input) plan which produces symbols: [A, B, C]
+ *   - (subquery) Aggregation(GROUP BY 1; functions: [sum(F), count(), ...]
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: This should say GROUP BY ()

File: presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java]: done

**Comment by kokosing** [comment]
@martint 
I fixed the commit message. Previously I have two commits one for `count` and second for other aggregate functions. Then I merged them, somehow I failed to merged commit messages. Sorry about that.

The query you posted is somehow tricky. `EXISTS` subquery ends with double (nested) aggregation. Such query is not yet supported, though it unveiled the bug. `decorrelateFilters` could return still correlated node. I added a check (and tests) for that. 

I will try to extend the support for correlated subqueries to handle queries the one you posted.

**Comment by kokosing** [comment]
@martint Rebased and all the comments addresssed. Superseded by #6190.

**Pull request closed without merging**